### DISCUSSION TITLE
Host HID_controller issues with controllers on RP2040

### DISCUSSION THREAD

**Discussion by dshadoff** [hathach/tinyusb#1052] [General]
### Operating System

Others

### Board

Raspberry Pi Pico (RP2040)

### Firmware

examples/host/hid_controller
Cloned from current tinyusb 'head' on August 25
Used in conjunction with current release of pico-sdk (1.2.0)

### What happened ?

Tested with 2 controllers; results are as follows:

1) Hori HORIPAD mini4 was enumerated/identified  as VID = 0f0d, PID = 00ee, but example did not respond to any controller actions or button presses
-> Modifying src/hid_app.c ( function 'is_sony_ds4()' ) to compare against the HORI HORIPAD mini4's VID/PID (0f0d & 00ee as mentioned above) solved the issue, and compatibility was confirmed.  Perhaps this VID/PID explicit check is not the best design for this example; perhaps a compatibility warning in the log output or an alternate 'minimal' implementation (if possible) would be more helpful.  At least a message in the output to tell the user that no output should be expected if the  controller is not in a compatibility list.

2) 8BitDo SN30 pro was not initially enumerated, and did not respond to any controller actions or button presses.
PANIC was seen when used with a 'LOG=3' version of the example code.

### How to reproduce ?

(For 8BitDo controller failure)
Run example after loading into the microcontroller board
The "TinyUSB Host HID Controller Example" title appears, and after about 5 seconds, "PANIC" appears.
Even when compiled with LOG=3, no useful log information was available.

### Debug Log

[8bitdo_log3.log](https://github.com/hathach/tinyusb/files/7050874/8bitdo_log3.log)

**Comment by hathach** [comment]
moved to discussion since this is not bug, it is how the code is written to do. the hid controller example code say clearly it will only work with PS4 dualshock. Like other hid, to universally work with all devices, you need to either
1. write an universal hid parser to know the report layout of x,y,z, hat, button etc.. unline keyboard/mouse there is no boot gamepad
2. known report layout of specific device using vid/pid and extract info like with PS4 dualshock.

The code is just an example for how to do the later. If you want to make it work with other specific device, you have to add your own code to handle it and submit PR if you want to help others. I only have PS4 controller and have no plan to add anything else.

**Reply by dshadoff** [reply]
For the Hori, which was recognized, I accept this (although the log message could be more explicit about limitations).

But the "8bitdo SN30 Pro" received a PANIC before the VID/PID was identified and printed; this appears to be a bug.  I don't believe that custom parse code would be possible in this case, since it is never reaches the identification stage.  Please review the log.

**Reply by hathach** [reply]
can you provide the config & hid descriptor for the device for analysis

**Reply by dshadoff** [reply]
Apparently this controller does not get mounted on RPi Linux.
While I have another Linux device which successfully uses this controller (MiSTer), it also isn't able to gather the information requested, as the extract executables show errors.  Not sure why the controller works without issues under those conditions, but I won't question it.

So, I will abandon this inquiry for that controller.
I will create a new conversation if/when I find a device with issues, where the config & hid information can be gathered more cleanly.

**Reply by hathach** [reply]
thanks for the update, sound like issue with the device itself. Maybe it expect a certain command in enumeration sequence. It is hard to know, if you really want to use it. Try to analyze the enumeration on platform that it works and try to reverse engineer from that command by command.
